File: backend/api/brain/ledger.js
/**
 * Brain ledger helpers — shared write path for all ingestion sources.
 * Every external event writes a LedgerEvent first, then derives graph state.
 */

const { getPrisma } = require('../utils/prisma');
const { checkSelfIdentity } = require('./selfIdentity');

function canonicalName(name) {
  return String(name || '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

// Order-insensitive deep equality via stable-key JSON. Used so payload-merge
// backfills don't rewrite a row whose values are unchanged but whose stored
// key order differs from a freshly-built object.
function stableStringify(v) {
  if (v === null || typeof v !== 'object') return JSON.stringify(v);
  if (Array.isArray(v)) return `[${v.map(stableStringify).join(',')}]`;
  return `{${Object.keys(v).sort().map((k) => `${JSON.stringify(k)}:${stableStringify(v[k])}`).join(',')}}`;
}

/**
 * Write an immutable ledger event. Returns the created row.
 * All ingestion paths must call this before writing any BrainAssertion or BrainInboxItem.
 */
async function writeLedgerEvent({
  eventType,
  schemaVersion = 1,
  occurredAt,
  source,
  sourceId = null,
  actorType = null,
  actorId = null,
  payload,
  updatePayload = false, // when an event already exists, merge `payload` into it
                         // instead of skipping. Used by enrichment backfills that
                         // add new fields (e.g. identitySignals, customerEntityId)
                         // to events written by an earlier sync. New keys win.
}) {
  const prisma = getPrisma();
  if (sourceId) {
    const existing = await prisma.ledgerEvent.findFirst({
      where: {
        eventType,
        source,
        sourceId: String(sourceId),
        tombstonedAt: null,
      },
      orderBy: { createdAt: 'desc' },
    });
    if (existing) {
      if (updatePayload && payload && typeof payload === 'object') {
        const merged = { ...(existing.payload || {}), ...payload };
        // Only write if something actually changed — keeps re-runs no-op.
        // Order-insensitive compare: stored key order differs from a fresh build.
        if (stableStringify(merged) !== stableStringify(existing.payload || {})) {
          const updated = await prisma.ledgerEvent.update({
            where: { id: existing.id },
            data: { payload: merged },
          });
          return { ...updated, _existing: true, _updated: true };
        }
      }
      return { ...existing, _existing: true };
    }
  }
  try {
    return await prisma.ledgerEvent.create({
      data: {
        eventType,
        schemaVersion,
        occurredAt: occurredAt ? new Date(occurredAt) : new Date(),
        source,
        sourceId: sourceId ? String(sourceId) : null,
        actorType,
        actorId: actorId ? String(actorId) : null,
        payload,
      },
    });
  } catch (err) {
    // A concurrent writer may have inserted the same (eventType, source, sourceId)
    // between the findFirst above and this create. The partial unique index
    // (20260627000100_brain_unique_constraints) now enforces this; treat the
    // race as a dedupe hit rather than an error.
    if (err?.code === 'P2002' && sourceId) {
      const existing = await prisma.ledgerEvent.findFirst({
        where: { eventType, source, sourceId: String(sourceId), tombstonedAt: null },
        orderBy: { createdAt: 'desc' },
      });
      if (existing) return { ...existing, _existing: true };
    }
    throw err;
  }
}

async function findOrCreateEntity({ entityType, name, properties = null, status = 'active' }) {
  const prisma = getPrisma();

  // Never mint the business's own identity as a counterparty (Customer/Vendor/
  // Person). This is the single choke point for triage, intake, constraint, and
  // inbox ingestion — see backend/api/brain/selfIdentity.js. Existing rows that
  // match are still returned (so callers can re-link), but new ones are refused.
  const self = checkSelfIdentity(entityType, name);
  if (self.blocked) {
    const existingSelf = await prisma.brainEntity.findFirst({
      where: {
        entityType,
        tombstonedAt: null,
        OR: [
          { canonicalName: canonicalName(name) },
          { name: { equals: name, mode: 'insensitive' } },
        ],
      },
    });
    return { entity: existingSelf || null, created: false, blocked: true, blockReason: self.reason };
  }

  const normalized = canonicalName(name);
  const existing = await prisma.brainEntity.findFirst({
    where: {
      entityType,
      tombstonedAt: null,
      OR: [
        { canonicalName: normalized },
        { name: { equals: name, mode: 'insensitive' } },
        { aliases: { some: { alias: { equals: name, mode: 'insensitive' } } } },
      ],
    },
  });
  if (existing) return { entity: existing, created: false };

  const entity = await prisma.brainEntity.create({
    data: {
      entityType,
      name,
      canonicalName: normalized,
      properties,
      status,
    },
  });
  return { entity, created: true };
}

/**
 * Find a Vendor BrainEntity by merchant name (exact or alias match).
 * Returns the entity id or null.
 */
async function findVendorByName(merchantName) {
  const prisma = getPrisma();
  if (!merchantName) return null;

  const normalized = merchantName.toLowerCase().trim();

  // Direct name match (case-insensitive)
  const direct = await prisma.brainEntity.findFirst({
    where: {
      entityType: 'Vendor',
      status: { not: 'tombstoned' },
      name: { equals: merchantName, mode: 'insensitive' },
    },
    select: { id: true },
  });
  if (direct) return direct.id;

  // Alias match
  const alias = await prisma.brainEntityAlias.findFirst({
    where: {
      alias: { equals: merchantName, mode: 'insensitive' },
      entity: { entityType: 'Vendor', status: { not: 'tombstoned' } },
    },
    select: { entityId: true },
  });
  if (alias) return alias.entityId;

  // Normalized alias match
  const normalizedAlias = await prisma.brainEntityAlias.findFirst({
    where: {
      alias: { contains: normalized, mode: 'insensitive' },
      entity: { entityType: 'Vendor', status: { not: 'tombstoned' } },
    },
    select: { entityId: true },
  });
  return normalizedAlias?.entityId ?? null;
}

/**
 * Upsert a BrainInboxItem from a ledger event.
 * Used when an event needs human triage rather than automatic graph update.
 */
async function createInboxItem({ rawContent, source, attachments = null, ledgerEventId = null }) {
  const prisma = getPrisma();
  return prisma.brainInboxItem.create({
    data: {
      rawContent,
      source,
      attachments: attachments ?? undefined,
      capturedAt: new Date(),
      status: 'pending',
    },
  });
}

module.exports = { writeLedgerEvent, findVendorByName, createInboxItem, findOrCreateEntity, canonicalName };

File: brain-sidecar/ledger.py
"""
Write-path helpers — mirrors backend/api/brain/ledger.js.
Every extraction job must call write_ledger_event() before writing assertions.
"""

import uuid
from datetime import datetime, timezone
from db import execute, query


def canonical_name(name: str) -> str:
    """Stable lookup key for entity names."""
    import re
    return re.sub(r'\s+', ' ', re.sub(r'[^a-z0-9]+', ' ', (name or '').lower())).strip()


def write_ledger_event(
    event_type: str,
    source: str,
    payload: dict,
    occurred_at=None,
    source_id=None,
    actor_type: str = 'system',
    schema_version: int = 1,
) -> str:
    """Insert a LedgerEvent row. Returns the new id."""
    if source_id:
        existing = query(
            """
            SELECT id FROM "LedgerEvent"
            WHERE "eventType" = %s AND source = %s AND "sourceId" = %s
              AND "tombstonedAt" IS NULL
            ORDER BY "createdAt" DESC
            LIMIT 1
            """,
            (event_type, source, source_id),
        )
        if existing:
            return existing[0]['id']

    eid = str(uuid.uuid4())
    ts = (occurred_at or datetime.now(timezone.utc)).isoformat()
    import json
    execute(
        """
        INSERT INTO "LedgerEvent"
            (id, "eventType", "schemaVersion", "occurredAt", source, "sourceId",
             "actorType", payload, "createdAt")
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s::jsonb, NOW())
        """,
        (eid, event_type, schema_version, ts, source, source_id,
         actor_type, json.dumps(payload)),
    )
    return eid


def write_assertion(
    src_id: str,
    dst_id: str,
    rel_type: str,
    ledger_event_id: str,
    confidence: float = 0.8,
    metadata=None,
    valid_from=None,
    provisional: bool = True,
    created_by: str = 'python_extractor',
) -> str:
    """Insert a BrainAssertion row. Returns the new id."""
    import json
    rel = rel_type.upper()
    if ledger_event_id:
        existing = query(
            """
            SELECT id FROM "BrainAssertion"
            WHERE "srcId" = %s AND "dstId" = %s AND "relType" = %s
              AND "sourceId" = %s AND "retractedAt" IS NULL
            ORDER BY "createdAt" DESC
            LIMIT 1
            """,
            (src_id, dst_id, rel, ledger_event_id),
        )
        if existing:
            return existing[0]['id']

    aid = str(uuid.uuid4())
    ts = (valid_from or datetime.now(timezone.utc)).isoformat()
    execute(
        """
        INSERT INTO "BrainAssertion"
            (id, "srcId", "dstId", "relType", metadata, "validFrom", "knownFrom",
             confidence, "sourceType", "sourceId", "createdBy", provisional, "createdAt")
        VALUES (%s, %s, %s, %s, %s::jsonb, %s, NOW(), %s, 'python_extractor', %s, %s, %s, NOW())
        """,
        (aid, src_id, dst_id, rel,
         json.dumps(metadata or {}), ts,
         confidence, ledger_event_id, created_by, provisional),
    )
    return aid


def find_or_create_entity(entity_type: str, name: str) -> tuple[str, bool]:
    """
    Look up an entity by type+name (case-insensitive).
    Creates it if missing. Returns (entity_id, created).
    """
    norm = canonical_name(name)
    import json
    from db import get_conn
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute('SELECT pg_advisory_xact_lock(hashtext(%s))', (f'{entity_type}:{norm}',))
            cur.execute(
                """
                SELECT id FROM "BrainEntity"
                WHERE "entityType" = %s
                  AND COALESCE("canonicalName", LOWER(name)) = %s
                  AND "tombstonedAt" IS NULL
                LIMIT 1
                """,
                (entity_type, norm),
            )
            row = cur.fetchone()
            if row:
                conn.rollback()
                return row['id'], False

            cur.execute(
                """
                SELECT ea."entityId" FROM "BrainEntityAlias" ea
                JOIN "BrainEntity" e ON e.id = ea."entityId"
                WHERE (
                    LOWER(ea.alias) = LOWER(%s)
                    OR COALESCE(e."canonicalName", LOWER(e.name)) = %s
                  )
                  AND e."entityType" = %s
                  AND e."tombstonedAt" IS NULL
                LIMIT 1
                """,
                (name, norm, entity_type),
            )
            row = cur.fetchone()
            if row:
                conn.rollback()
                return row['entityId'], False

            eid = str(uuid.uuid4())
            cur.execute(
                """
                INSERT INTO "BrainEntity"
                    (id, "entityType", name, "canonicalName", status, properties, "createdAt", "updatedAt")
                VALUES (%s, %s, %s, %s, 'active', %s::jsonb, NOW(), NOW())
                """,
                (eid, entity_type, name, norm, json.dumps({})),
            )
        conn.commit()
        return eid, True
    except Exception:
        conn.rollback()
        raise


def already_processed(source: str, source_id: str) -> bool:
    """Return True if a LedgerEvent with this source+sourceId already exists."""
    rows = query(
        'SELECT id FROM "LedgerEvent" WHERE source = %s AND "sourceId" = %s LIMIT 1',
        (source, source_id),
    )
    return len(rows) > 0


def bulk_already_processed(source: str, source_ids: list) -> set:
    """Return the set of source_ids that have already been processed."""
    if not source_ids:
        return set()
    placeholders = ','.join(['%s'] * len(source_ids))
    rows = query(
        f'SELECT "sourceId" FROM "LedgerEvent" WHERE source = %s AND "sourceId" IN ({placeholders})',
        (source, *source_ids),
    )
    return {r['sourceId'] for r in rows}
